src/darwin/Framework/CHIP/MTRBaseDevice.h

```diff
@@ -302,6 +302,30 @@ extern NSString * const MTRArrayValueType;
                                       completion:(MTRDeviceOpenCommissioningWindowHandler)completion
     API_AVAILABLE(ios(16.2), macos(13.1), watchos(9.2), tvos(16.2));
 
+/**
+ * Reads events from the device.
+ *
+ * Nil values for endpointID, clusterID, eventID indicate wildcards
+ * (e.g. nil eventID means "read all the events from the endpoint(s) and
+ * cluster(s) that match endpointID/clusterID").
+ *
+ * If all of endpointID, clusterID, eventID are non-nil, a single
+ * event will be read.
+ *
+ * If all of endpointID, clusterID, eventID are nil, all events on the
+ * device will be read.
+ *
+ * A non-nil eventID along with a nil clusterID will only succeed if the
+ * event ID is for a global event that applies to all clusters.
+ */
+
+- (void)readEventsWithEndpointID:(NSNumber * _Nullable)endpointID
```

Comment:
Why do we want to add this API?  What are the use cases?

Reply:
We are preparing Matter Certification for a Matter controller iOS app which can use this API to pass test cases related to "read event".

Reply:
@seokhee-lee Does the app actually "read events" one at a time like this (as opposed to subscribing to events)?  Or is this just so it can pass the relevant certification tests?  I ask, because those tests shoud be conditioned on PICS that say whether the app actually does this....

Reply:
@seokhee-lee I'm just trying to understand what the use cases are for "read event", mostly.  But if you're planning to do it, we can certainly add API for it in the framework; we just need to get rid of the code duplication to make this more maintainable.

Reply:
@bzbarsky-apple There are several "read event" test cases which we are considering for certification. One of the test cases is DRLK-3.3 in 22-27353-001_Matter-1.0-Application-Cluster-TestPlans which can give some information to understand what the use cases are for "read event". We are also reviewing the codes to be more maintainable and will update them if there are some modifications to enhance it.

Reply:
Thank you for the pointer to that test plan.  That test plan .... does not use "read event" as far as I can see.  As best I can tell, it's assuming (but not stating) that the client has a subscription to the server, so when the server generates events they are reported to the client...

Reply:
In our experience (we got the controller certifications for webOS and Android), tester requested logs of read event results of server (mandatory) and client (optional) for the test case. Unlike read attribute, read event returns all history of the events but not only the value at that point in time.

Reply:
@cjandhyala could you please take a look?  That does _not_ seem like a reasonable requirement.

Reply:
@seokhee-lee you don't need readEvent per say, The purpose of TC is the client able to receive the events, on the Darwin platform you can subscript to all events ('subscribe-all-events ')  and filter for the StartUp and other events you are interested to look for. On the example app (TH) you can trigger events using the commands mentioned at https://github.com/project-chip/connectedhomeip/blob/e56c2ce42f4bf5d7a5aab8898174b4f14e7888e6/examples/all-clusters-app/linux/README.md

For DUT as client TCs, The chip-tool commands provided in the verification steps doc are for example on how the same functionality verified using chip-tool. DUTs can run any equivalent command that verifies the purpose of the test.

Reply:
@bzbarsky-apple , @cjandhyala , thanks for your supports. I think that there is some misunderstanding in our side. I will discuss more with our app development team on whether they have the other requirements except for certification. If it was just for certification, I will close this pull request.

Reply:
Just to be clear: this is totally worth adding if people are going to use it.  We just want to share the code with the attribute read setup as much as possible.

Reply:
@bzbarsky-apple I checked the requirement on read/subscribe event API from our app team. They said that "subscribing all event" has some performance issues and so our iOS Matter Controller app will read/subscribe some selective events by using the proposed APIs. I am not sure whether this reason is enough or not in open source side. If the reason is not enough to be merged in upstream, We will manage this patch in our local build system.

Reply:
> so our iOS Matter Controller app will read/subscribe some selective events by using the proposed APIs

OK. That seems fine, as long as you realize the proposed API will in practice only allow you to subscribe to three event paths reliably (that's how many subscriptions a device is required to support per fabric).

Given that, I think the only outstanding issue is the code duplication compared to the "subscribe attribute" bit.

Reply:
@bzbarsky-apple "subscribe to one event" already has been implemented in our side and we will add "subscribe to one event" in this PR soon. Now, attribute in darwin matter framework also seems not to support "read/subscribe selected ones". I think that it can make app developing more convenient. We will try to do PR for it after this PR.

Reply:
@bzbarsky-apple "subscribe to one event" is added in this PR. Please review again and let me know if more modifications are needed.